File: app/src/main/java/io/github/rsookram/txt/reader/ContentLoader.java
package io.github.rsookram.txt.reader;

import android.content.ContentResolver;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.github.rsookram.txt.Book;
import io.github.rsookram.txt.BookContent;
import io.github.rsookram.txt.Line;

public class ContentLoader {

    private final ContentResolver contentResolver;

    public ContentLoader(ContentResolver contentResolver) {
        this.contentResolver = contentResolver;
    }

    public BookContent load(Book book) {
        List<Line> lines = getLines(book);
        Line lastLine = lines.isEmpty() ? null : lines.get(lines.size() - 1);
        int length = lastLine != null ? lastLine.offset + lastLine.text.length() : 0;
        return new BookContent(lines, length);
    }

    private List<Line> getLines(Book book) {
        InputStream stream;
        try {
            stream = contentResolver.openInputStream(book.uri);
        } catch (FileNotFoundException e) {
            return Collections.emptyList();
        }

        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(stream));

        List<Line> lines = new ArrayList<>();

        try {
            bufferedReader.lines()
                    .map(String::trim)
                    .forEach(str -> {
                        Line prevLine = lines.isEmpty() ? null : lines.get(lines.size() - 1);
                        int offset = prevLine != null ? prevLine.offset + prevLine.text.length() : 0;
                        lines.add(new Line(str, offset));
                    });
        } catch (Exception e) {
            return Collections.emptyList();
        }

        return lines;
    }
}

File: app/src/main/java/io/github/rsookram/txt/reader/ContentLoader.kt
package io.github.rsookram.txt.reader

import android.content.ContentResolver
import io.github.rsookram.txt.Book
import io.github.rsookram.txt.BookContent
import io.github.rsookram.txt.Line

class ContentLoader(private val contentResolver: ContentResolver) {

    fun load(book: Book): BookContent {
        val lines = getLines(book)
        val lastLine = lines.lastOrNull()
        val length = if (lastLine != null) lastLine.offset + lastLine.text.length else 0
        return BookContent(lines, length)
    }

    private fun getLines(book: Book): List<Line> {
        val lineStrings = getLineStrings(book)

        val lines = mutableListOf<Line>()
        var offset = 0

        lineStrings.forEach { str ->
            lines.add(Line(str, offset))
            offset += str.length
        }

        return lines
    }

    private fun getLineStrings(book: Book): Sequence<String> {
        val stream = contentResolver.runCatching { openInputStream(book.uri) }
            .getOrNull()
            ?: return emptySequence()

        return stream
            .bufferedReader()
            .lineSequence()
            .map(String::trim)
    }
}
